Create Dockerfile for installing storage backends for running kronos unit tests.

lydiagu opened the pull request:
@silasbw I'm not sure about how to choose between creating runit entries in Phusion and just adding scripts to `/etc/my_init.d`. I went with the later because running cassandra in the foreground causes logs to go to stdout, which was overwhelming.

lydiagu commented:
```
test_delete (test_apis.TestKronosAPIs) ... ok
test_get (test_apis.TestKronosAPIs) ... ok
test_infer_schema (test_apis.TestKronosAPIs) ... ok
test_namespaces (test_apis.TestKronosAPIs) ... ok
test_put (test_apis.TestKronosAPIs) ... ok
test_streams (test_apis.TestKronosAPIs) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.498s

OK
> [memory]: common
test_delete (test_apis.TestKronosAPIs) ... ok
test_get (test_apis.TestKronosAPIs) ... FAIL
test_infer_schema (test_apis.TestKronosAPIs) ... ok
test_namespaces (test_apis.TestKronosAPIs) ... ok
test_put (test_apis.TestKronosAPIs) ... ok
test_streams (test_apis.TestKronosAPIs) ... ok

======================================================================
FAIL: test_get (test_apis.TestKronosAPIs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/chronology/kronos/tests/common/test_apis.py", line 136, in test_get
    self.assertEqual(len(events), 4)
AssertionError: 2 != 4

----------------------------------------------------------------------
Ran 6 tests in 1.312s

FAILED (failures=1)
> [cassandra]: common
test_delete (test_apis.TestKronosAPIs) ... ok
test_get (test_apis.TestKronosAPIs) ... ok
test_infer_schema (test_apis.TestKronosAPIs) ... ok
test_namespaces (test_apis.TestKronosAPIs) ... ok
test_put (test_apis.TestKronosAPIs) ... ok
test_streams (test_apis.TestKronosAPIs) ... ok

----------------------------------------------------------------------
Ran 6 tests in 2.984s

OK
> [elasticsearch]: common
test_index (test_cassandra.TestCassandraBackend)
Inserts events in to a stream and ensures that the index column ... ok
test_overlapping_shards (test_cassandra.TestCassandraBackend)
Tests that changing bucket widths doesn't break shit. First inserts ... FAIL
test_stream_sharding (test_cassandra.TestCassandraBackend)
Inserts elements that fall into the same time bucket and checks if they ... ok
test_stream_splitting (test_cassandra.TestCassandraBackend)
Inserts events that fall into multiple time buckets and ensures that ... FAIL

======================================================================
FAIL: test_overlapping_shards (test_cassandra.TestCassandraBackend)
Tests that changing bucket widths doesn't break shit. First inserts
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/chronology/kronos/tests/storage/cassandra/test_cassandra.py", line 153, in test_overlapping_shards
    self.assertEqual(len(events), 120)
AssertionError: 119 != 120

======================================================================
FAIL: test_stream_splitting (test_cassandra.TestCassandraBackend)
Inserts events that fall into multiple time buckets and ensures that
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/chronology/kronos/tests/storage/cassandra/test_cassandra.py", line 93, in test_stream_splitting
    self.assertEqual(len(events), 20)
AssertionError: 19 != 20

----------------------------------------------------------------------
Ran 4 tests in 1.803s

FAILED (failures=2)
> [cassandra]: cassandra
test_endpoints (test_serving_mode.TestServingMode) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.015s

OK
> [memory]: conf
test_aliasing (test_elasticsearch.TestElasticSearchBackend) ... ok
test_concurrent_rollover (test_elasticsearch.TestElasticSearchBackend) ... ok
test_rollover (test_elasticsearch.TestElasticSearchBackend) ... ok

----------------------------------------------------------------------
Ran 3 tests in 4.684s

OK
> [elasticsearch]: elasticsearch
```

marcua commented on kronos/tests/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+  dsc20=2.0.11-1 \
+  cassandra=2.0.11 \
+  elasticsearch
+RUN service cassandra stop
```
why do you have to do this?

lydiagu replied:
I thought I had to from http://docs.datastax.com/en/cassandra/2.0/cassandra/install/installDeb_t.html, but it was just to clear the data directory, which doesn't matter for us.

marcua commented on kronos/tests/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+  python-all-dev \
+  python-pip \
+  python2.7 \
+  supervisor \
```
needed?

lydiagu replied:
removed

marcua commented on kronos/tests/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# https://docs.docker.com/articles/dockerfile_best-practices/
+# Check for updates:
+#   https://github.com/phusion/baseimage-docker/blob/master/Changelog.md
+FROM phusion/baseimage:0.9.16
```
can you build onthe kronos one?

lydiagu replied:
I could.. but I thought it would be less confusing for the user to be able to test the current code, since the kronos image might not be the hash you want.

marcua commented on kronos/tests/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# Dockerfile for running kronos tests with various backends.
```
high-level design question: should this file actually install Kronos, or should it only contain the storage engines we need, and Kronos will link the storage engines in to communicate with them?

lydiagu replied:
You're right, no need to install. Still copying the code in though, are you saying we shouldn't copy the entire repo either?

marcua commented:
The Cassandra test failure is a known problem likely due to Cassandra's consistency model (#205).  No need to solve them in this diff.

Finished the review---more design questions than suggestions at this point.

usmanm commented:
The prints seem to be re-ordered. `> [<backend>]: <test>` is being printed after the output of the test. Weird.

lydiagu commented:
Changed from running cassandra/elasticsearch as services.
Added setup for elasticsearch because running from /usr/share/elasticsearch seems to not find the config files in /etc/elasticsearch. I followed recommendation in http://elasticsearch-users.115913.n3.nabble.com/Elasticsearch-throwing-exception-while-starting-on-Redhat-td4061067.html. I guess it looks in /usr/share/elasticsearch/config for config files?

marcua commented:
LGTM.  I bet once we hook this up to TravisCI or something, we'll need to link the latest codebase in as a volume rather than COPY it in, but we can figure that out when we get there.  Anyone else opposed to merging?
